tests/test_bigquery.py

```diff
@@ -1,38 +1,67 @@
 import unittest
 import os
 import threading
+from test.support import EnvironmentVarGuard
 
 from http.server import BaseHTTPRequestHandler, HTTPServer
 
 from google.cloud import bigquery
+from google.auth.exceptions import DefaultCredentialsError
 
 HOSTNAME = "127.0.0.1"
 PORT = 8000
 URL = "http://%s:%s" % (HOSTNAME, PORT)
 
 class TestBigQuery(unittest.TestCase):
-    def test_proxy(self):
-        httpd = HTTPServer((HOSTNAME, PORT), HTTPHandler)
-        threading.Thread(target=httpd.serve_forever).start()
-        client = bigquery.Client()
+    def _test_proxy(self, client, should_use_proxy):
+        class HTTPHandler(BaseHTTPRequestHandler):
```

Comment:
Can this be left at the module-level for better readability?  I don't see any currying taking place but I might be missing it.

Reply:
Yeah, I spent some time trying to do this properly but the issue is that HTTPHandler.called is a class variable, so if it is at the module level, the first time it's set to true from one test running, other tests will see it as true (in fact this was happening in my tests at first). Ideally 'called' would be an instance variable but HTTPServer/TCPServer never actually keep around the instance of HTTPHandler (they call RequestHandlerClass internally) for us to be able to inspect. So this is kind of a hack to keep the class variable scoped on a per test basis.

Reply:
You could reset the class attribute in the test's setup method, or here?  Tests don't run in parallel and never will.

You can also define a base class module-level, then subclass that within this function with `called = False`.

Reply:
Yeah what you do is fine if you prefer it.  Tests always run sequentially in google3 (parallelism is provided at a coarser level, not in-process), which is why I'd prefer the above for clarity / avoiding too much nesting.  Sorry for the subclass suggestion, that won't work indeed.

Having a fake BQ server would probably help too, but none of that is provided for GCP APIs.